# Task 8 Jupyter Notebook


Create a notebook called weather.ipynb at the root of your repository. In this notebook, write a brief report explaining how you completed Tasks 1 to 7. Provide short descriptions of the commands used in each task and explain their role in completing the tasks.

## Task 1: Create Directory Structure


**Task** Using the command line, create a directory (that is, a folder) named data at the root of your repository. Inside data, create two subdirectories: timestamps and weather.

In [288]:
%%bash

mkdir -p data


### Define directory paths


In [289]:
%%bash
mkdir -p data/timestamps data/weather

## Task 2: Append Current Date and Time to now.txt


**Task:**
Navigate to the data/timestamps directory. Use the date command to output the current date and time, appending the output to a file named now.txt. Make sure to use the >> operator to append (not overwrite) the file. Repeat this step ten times, then use the more command to verify that now.txt has the expected content.

In [290]:
%%bash

cd data/timestamps

for i in {1..10}; do
  date >> now.txt
done

more now.txt

::::::::::::::
now.txt
::::::::::::::
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:52:55.631722
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 21:55:36.943599
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-11 22:02:23.766387
2024-12-15 21:06:04.731326
2024-12-15 21:06:04.731326
2024-12-15 21:06:04.731326
2024-12-15 21:06:04.731326
2024-12-15 21:06:04.731326
2024-12-15 21:06:

## Task 3: Append Formatted Timestamps to formatted.txt


**Task:**
Run the date command again, but this time format the output using YYYYmmdd_HHMMSS (e.g., 20261114_130003 for 1:00:03 PM on November 14, 2026). Refer to the date man page (using man date) for more formatting options. (Press q to exit the man page). Append the formatted output to a file named formatted.txt.

In [291]:
%%bash

date "+%Y%m%d_%H%M%S" >> data/timestamps/formatted.txt



## Task 4: Create Timestamped Files


**Task:**
Use the touch command to create an empty file with a name in the YYYYmmdd_HHMMSS.txt format. You can achieve this by embedding your date command in backticks ` into the touch command. You should no longer use redirection (>>) in this step.

In [292]:
%%bash

timestamps_dir="data/timestamps"

touch "$timestamps_dir/$(date "+%Y%m%d_%H%M%S").txt"

## Task 5: Download Today’s Weather Data


**Task:**
Change to the data/weather directory. Download the latest weather data for the Athenry weather station from Met Eireann using wget. Use the -O <filename> option to save the file as weather.json. The data can be found at this URL:
https://prodapi.metweb.ie/observations/athenry/today.


In [293]:
%%bash
cd data/weather

wget -O weather.json https://prodapi.metweb.ie/observations/athenry/today

--2024-12-20 03:52:54--  https://prodapi.metweb.ie/observations/athenry/today
Resolving prodapi.metweb.ie (prodapi.metweb.ie)... 139.162.185.91
Connecting to prodapi.metweb.ie (prodapi.metweb.ie)|139.162.185.91|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [application/json]
Saving to: ‘weather.json’

     0K .                                                       335M=0s

2024-12-20 03:52:55 (335 MB/s) - ‘weather.json’ saved [1234]



## Task 6: Save Weather Data with Timestamped Filenames


**Task:**
Modify the command from Task 5 to save the downloaded file with a timestamped name in the format YYYYmmdd_HHMMSS.json.

In [294]:
%%bash


cd data/timestamps

# Use the curl command to download the file and save it with a timestamped name
curl -o "$(date "+%Y%m%d_%H%M%S").json" "http://example.com/path/to/file.json"

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1256  100  1256    0     0   4396      0 --:--:-- --:--:-- --:--:--  4391


## Task 7: Automate the Process with weather.sh

**Task description:** 
Write a bash script called `weather.sh` in the root of your repository. This script should automate the process from Task 6, saving the weather data to the `data/weather` directory. Make the script executable and test it by running it.

In [295]:
%%bash

cd data/weather

curl -o $(date +"%Y%m%d_%H%M%S").json https://prodapi.metweb.ie/observations/athenry/today




  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1234    0  1234    0     0   1446      0 --:--:-- --:--:-- --:--:--  1444


In [296]:
%%bash

chmod +x weather.sh

In [297]:
%%bash

./weather.sh

./weather.sh: line 1: fg: no job control
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1234    0  1234    0     0   1444      0 --:--:-- --:--:-- --:--:--  1443


In [298]:
%%bash

ls data/weather

20241211_194823.json
20241211_213106.json
20241211_215427.json
20241211_215538.json
20241211_220225.json
20241215_210607.json
20241218_234319.json
20241218_234614.json
20241219_012923.json
20241219_013443.json
20241219_014445.json
20241219_014452.json
20241219_015016.json
20241219_020828.json
20241219_232349.json
20241219_234817.json
20241220_015823.json
20241220_015825.json
20241220_023309.json
20241220_023409.json
20241220_033333.json
20241220_034027.json
20241220_034654.json
20241220_034656.json
20241220_035256.json
20241220_035257.json
weather.json


## Task 8: Notebook
**Task description:** 
Create a notebook called weather.ipynb at the root of your repository. In this notebook, write a brief report explaining how you completed Tasks 1 to 7. Provide short descriptions of the commands used in each task and explain their role in completing the tasks.

Using the "touch" command, I had created an empyty jupyter notebook, and using the "ls" command I had confirmed that the file had been created successfully:



In [299]:
%%bash

touch weather.ipynb


In [300]:
%%bash

ls

README.md
data
weather.ipynb
weather.sh


## Task 9: pandas
**Task description:** 
In your weather.ipynb notebook, use the pandas function read_json() to load in any one of the weather data files you have downloaded with your script. Examine and summarize the data. Use the information provided "[data.gov.ie](https://data.gov.ie/dataset/todays-weather-athenry)" to write a short explanation of what the data set contains.

In [301]:
import pandas as pd


### Load a JSON file in the weather directory

In [302]:
df = pd.read_json('data/weather/20241218_234614.json')

In [303]:
df.head()

name  temperature symbol weatherDescription              text  \
0  Athenry           13    05n       Rain showers     "Rain shower"   
1  Athenry           13    05n       Rain showers     "Rain shower"   
2  Athenry           12    05n       Rain showers     "Rain shower"   
3  Athenry           12    05n       Rain showers     "Rain shower"   
4  Athenry           12    09n               Rain  "Moderate rain "   

   windSpeed windGust cardinalWindDirection  windDirection  humidity  \
0         22        -                    SW            225        96   
1         30        -                    SW            225        94   
2         22       44                    SW            225        96   
3         26       46                    SW            225        97   
4         28        -                    NW            315        95   

   rainfall  pressure    dayName       date reportTime  
0       0.3       993  Wednesday 2024-12-18      00:00  
1       0.7       992  Wednesday 2024-12-18      01:00  
2       0.8       991  Wednesday 2024-12-18      02:00  
3       0.8       991  Wednesday 2024-12-18      03:00  
4       1.6       993  Wednesday 2024-12-18      04:00

In [304]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   name                   24 non-null     object        
 1   temperature            24 non-null     int64         
 2   symbol                 24 non-null     object        
 3   weatherDescription     24 non-null     object        
 4   text                   24 non-null     object        
 5   windSpeed              24 non-null     int64         
 6   windGust               24 non-null     object        
 7   cardinalWindDirection  24 non-null     object        
 8   windDirection          24 non-null     int64         
 9   humidity               24 non-null     int64         
 10  rainfall               24 non-null     float64       
 11  pressure               24 non-null     int64         
 12  dayName                24 non-null     object        
 13  date   

In [305]:
df.isnull().sum()

name                     0
temperature              0
symbol                   0
weatherDescription       0
text                     0
windSpeed                0
windGust                 0
cardinalWindDirection    0
windDirection            0
humidity                 0
rainfall                 0
pressure                 0
dayName                  0
date                     0
reportTime               0
dtype: int64

In [306]:
df.describe()

temperature  windSpeed  windDirection   humidity   rainfall  \
count    24.000000  24.000000      24.000000  24.000000  24.000000   
mean      8.583333  13.875000     273.750000  89.708333   0.259583   
min       3.000000   6.000000     225.000000  83.000000   0.000000   
25%       7.000000   7.000000     258.750000  87.000000   0.000000   
50%       9.000000  11.000000     270.000000  88.000000   0.000000   
75%      10.000000  17.750000     315.000000  92.250000   0.150000   
max      13.000000  30.000000     315.000000  97.000000   1.900000   
std       2.796219   7.502536      34.898922   3.939534   0.528455   

          pressure                 date  
count    24.000000                   24  
mean   1000.875000  2024-12-18 00:00:00  
min     991.000000  2024-12-18 00:00:00  
25%     998.250000  2024-12-18 00:00:00  
50%    1003.000000  2024-12-18 00:00:00  
75%    1004.000000  2024-12-18 00:00:00  
max    1007.000000  2024-12-18 00:00:00  
std       5.160995                  NaN

## Summary of each data column:

**Temperature:** From 3°C to 13°C, with an average of 8.58°C.

**Wind Speed:** From 6 to 30 units/hour, with an average of 13.88.

**Wind Direction:** From 225° to 315°, with an average of 273.75°.

**Humidity:** Between 83% and 97%, with an average of 89.71%.

**Rainfall:** From 0.0 to 1.9 mm, averaging 0.26 mm.

**Pressure:** Between 991 and 1007 hPa, with an average of 1000.88 hPa.

**Date:** Observations were from the 18th of December 2024.

# References

1. Athenry weather station data: https://prodapi.metweb.ie/observations/athenry/today

2. Today's weather, Athenry, data.gov.ie: https://data.gov.ie/dataset/todays-weather-athenry

3. %%bash: https://www.geeksforgeeks.org/useful-ipython-magic-commands/

4. Create a directory: https://www.w3schools.com/python/ref_os_mkdir.asp

5. more command: https://www.geeksforgeeks.org/more-command-in-linux-with-examples/

6. wget vs curl: https://www.warp.dev/terminus/curl-vs-wget

7. chmod-x: https://www.geeksforgeeks.org/what-does-chmod-x-do-and-how-to-use-it/

8. JSON - Pandas: https://www.w3schools.com/python/pandas/pandas_json.asp